src/stats/rawsender.cpp

```diff
@@ -0,0 +1,169 @@
+        error = strprintf("Cannot get socket address for %s", m_host);
+        return;
+    }
+
+    SOCKET hSocket = ::socket(AF_INET, SOCK_DGRAM, IPPROTO_UDP);
+    if (hSocket == INVALID_SOCKET) {
+        error = strprintf("Cannot create socket (socket() returned error %s)", NetworkErrorString(WSAGetLastError()));
+        return;
+    }
+    m_sock = std::make_unique<Sock>(hSocket);
+
+    if (m_interval_ms == 0) {
+        LogPrintf("Send interval is zero, not starting RawSender queueing thread.\n");
+    } else {
+        m_interrupt.reset();
+        m_thread = std::thread(&util::TraceThread, "rawsender", [this] { QueueThreadMain(); });
+    }
+
+    LogPrintf("Started %sRawSender sending messages to %s:%d\n", m_thread.joinable() ? "threaded " : "",
+              m_host, m_port);
+}
+
+RawSender::~RawSender()
+{
+    // If there is a thread, interrupt and stop it
+    if (m_thread.joinable()) {
+        m_interrupt();
+        m_thread.join();
+    }
+    // Flush queue of uncommitted messages
+    QueueFlush();
+    // Reset connection socket
+    m_sock.reset();
+
+    LogPrintf("Stopped RawSender instance sending messages to %s:%d. %d successes, %d failures.\n",
+              m_host, m_port, m_successes, m_failures);
+}
+
+std::optional<std::string> RawSender::Send(const RawMessage& msg) noexcept
+{
+    // If there is a thread, append to queue
+    if (m_thread.joinable()) {
+        QueueAdd(msg);
+        return std::nullopt;
+    }
+    // There isn't a queue, send directly
+    return SendDirectly(msg);
+}
+
+std::optional<std::string> RawSender::SendDirectly(const RawMessage& msg) noexcept
+{
+    if (!m_sock) {
+        m_failures++;
+        return "Socket not initialized, cannot send message";
+    }
+
+    if (::sendto(m_sock->Get(), reinterpret_cast<const char*>(msg.data()),
+#ifdef WIN32
+                 static_cast<int>(msg.size()),
+#else
+                 msg.size(),
+#endif // WIN32
+                 /*flags=*/0, reinterpret_cast<struct sockaddr*>(&m_server.first), m_server.second) == SOCKET_ERROR) {
+        m_failures++;
+        return strprintf("Unable to send message to %s (sendto() returned error %s)", this->ToStringHostPort(),
+                         NetworkErrorString(WSAGetLastError()));
+    }
+
+    m_successes++;
+    return std::nullopt;
+}
+
+std::string RawSender::ToStringHostPort() const noexcept
+{
+    return strprintf("%s:%d", m_host, m_port);
+}
+
+void RawSender::QueueAdd(const RawMessage& msg) noexcept
+{
+    AssertLockNotHeld(cs);
+    LOCK(cs);
+
+    const auto& [batch_size, batch_delim] = m_batching_opts;
+    // If no batch size has been specified, simply add to queue
+    if (batch_size == 0) {
+        m_queue.push_back(msg);
+        return;
+    }
+
+    // We can batch, either create a new batch in queue or append to existing batch in queue
+    if (m_queue.empty() || m_queue.back().size() > batch_size) {
```

Comment:
```suggestion
    if (m_queue.empty() || m_queue.back().size() >= batch_size) {
```
because we add the message afterwards

Reply:
Resolved in latest push!